### ISSUE TITLE
Fix incorrect spacing handling in nolint directives

### ISSUE THREAD

**Issue by 796RCP92VZ** [golangci/golangci-lint#4561]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've read the typecheck section of the FAQ (https://golangci-lint.run/usage/faq/#why-do-you-have-typecheck-errors).
- [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

### Description of the problem

For nolint directive comments that have both leading whitespace and whitespace between `nolint:` and the linter list, the nolintlint linter gives a warning of the form “directive `// nolint: lll // ignore line length warnings` should be written without leading space as `//nolint: lll // ignore line length warnings`", whereas the correct directive comment would actually be `//nolint:lll // ignore line length warnings` (with no space between `nolint:` and `lll`). If the reader implements nolintlint's suggested fix in this case, `go fmt` will add a leading space (as of go1.19) because the format `//nolint: <comma-separated-linters>` does not conform to the [official Go directive format](https://go.dev/doc/comment#:~:text=A%20directive%20comment%20is%20a,blank%20lines%20in%20doc%20comments.), which is "a line matching the regular expression `//(line |extern |export |[a-z0-9]+:[a-z0-9])`", so `go fmt` thinks that an ordinary comment (as opposed to a directive) was intended and subsequently adds the space between "//" and the comment text that is required for an ordinary comment.

Additionally, nolint directive comments that do not have leading whitespace but do have whitespace between `nolint:` and the linter list, or do not have a linter list at all (i.e., just `//nolint` with no `:` thereafter, as opposed to the `//nolint:all` directive that was added to golangci-lint to address this issue after the release of go1.19) will not be marked with a warning by nolintlint at all, which also causes confusion due to non-compliance with `go fmt`. This is somewhat less likely to occur, since `go fmt` is often run automatically on save, and `go fmt` would add a space between `//` and `nolint` in these cases, but if someone saves a file without running `go fmt`, such a warning message would help them be compliant with `go fmt`'s expectations as well as the proper directive format in the future.

Related to #4544

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 1.54.2 built with go1.21.0 from 411e0bb on 2023-08-21T11:04:00Z
```

</details>

### Configuration

<details>

```console
linters:
  disable-all: true
  enable:
    - nolintlint
```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.22.1 darwin/arm64
GO111MODULE='on'
GOARCH='arm64'
GOBIN=''
GOCACHE='/Users/aaron.pradhan/Library/Caches/go-build'
GOENV='/Users/aaron.pradhan/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMODCACHE='/Users/aaron.pradhan/go/pkg/mod'
GOOS='darwin'
GOPATH='/Users/aaron.pradhan/go'
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/Users/aaron.pradhan/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.darwin-arm64'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/Users/aaron.pradhan/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.darwin-arm64/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.22.1'
GCCGO='gccgo'
AR='ar'
CC='clang'
CXX='clang++'
CGO_ENABLED='1'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/z4/_x30p3hj55qgzlpftpl23pnh0000gp/T/go-build2996606128=/tmp/go-build -gno-record-gcc-switches -fno-common'
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /Users/aaron.pradhan/work/garage /Users/aaron.pradhan/work /Users/aaron.pradhan /Users /] 
INFO [config_reader] Used config file .golangci.yaml 
INFO [lintersdb] Active 1 linters: [nolintlint]   
INFO [loader] Go packages loading at mode 575 (deps|exports_file|imports|types_sizes|compiled_files|files|name) took 401.246667ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 564.041µs 
INFO [linters_context/goanalysis] analyzers took 38.333µs with top 10 stages: nolintlint: 33.083µs, typecheck: 5.25µs 
INFO [runner] Issues before processing: 4, after processing: 1 
INFO [runner] Processors filtering stat (out/in): max_same_issues: 1/1, max_from_linter: 1/1, path_shortener: 1/1, severity-rules: 1/1, path_prettifier: 4/4, identifier_marker: 4/4, exclude: 4/4, filename_unadjuster: 4/4, source_code: 1/1, uniq_by_line: 1/1, sort_results: 1/1, cgo: 4/4, skip_dirs: 4/4, nolint: 1/4, diff: 1/1, max_per_file_from_linter: 1/1, fixer: 1/1, path_prefixer: 1/1, skip_files: 4/4, autogenerated_exclude: 4/4, exclude-rules: 4/4 
INFO [runner] processing took 522.709µs with stages: exclude-rules: 152.375µs, autogenerated_exclude: 142.042µs, nolint: 62.249µs, identifier_marker: 56.666µs, path_prettifier: 47.208µs, source_code: 40.791µs, skip_dirs: 16µs, uniq_by_line: 875ns, max_same_issues: 750ns, cgo: 708ns, path_shortener: 584ns, max_from_linter: 375ns, exclude: 333ns, filename_unadjuster: 292ns, fixer: 292ns, sort_results: 251ns, max_per_file_from_linter: 250ns, diff: 209ns, skip_files: 209ns, severity-rules: 166ns, path_prefixer: 84ns 
INFO [runner] linters took 139.781208ms with stages: goanalysis_metalinter: 139.213542ms 
dummy.go:3:1: directive `// nolint: lll // ignore line length warnings` should be written without leading space as `//nolint: lll // ignore line length warnings` (nolintlint)
// nolint: lll // ignore line length warnings
^
INFO File cache stats: 1 entries of total size 92B 
INFO Memory: 7 samples, avg is 34.8MB, max is 68.9MB 
INFO Execution took 555.593833ms
```

</details>

### A minimal reproducible example or link to a public repository

<details>

```go
// nolint: lll // ignore line length warnings
type MyInterface interface {}
```

</details>

### Validation

- [X] Yes, I've included all the information above (version, config, etc.).

**Comment by apradhan12** [comment]
@ldez Thanks for taking a look at this issue. Could you explain why you don't intend to fix this?

**Comment by oschwald** [comment]
The "additionally" part mentioned in the issue hit someone on my team today. They originally wrote `//nolint: revive` and running `golangci-lint` with `--fix` turned it into `// nolint: revive`, which made `nolintlint` fail. I have a vague memory of `nolintlint` or some other linter preventing the space before the linter name in the past, but maybe I am misremembering.

**Comment by ldez** [comment]
I put `won't fix` because:
1. Your issue wants to lead to the same thing I already refused with your PR #4544.
2. Some extra space cases are already fixed automatically, we have tests to prove that. (ex: `//  nolint:bob` -> `//nolint:bob`)
3. Your issue is a detail of a larger one (that we will not handle for now, intentionally, to avoid breaking changes)

The only thing we MAY do is create a `nolintlint` option to DETECT the invalid cases and create a report for them (a user should be able to enable/disable this behavior to avoid usage of `exclude` about them).

But this is not a "fix" for the incorrect spacing and this could not change the current behavior and the current supported syntax.

For me, your issue is off-topic by the goal it wants to achieve: "Fix incorrect spacing handling in nolint directives".

All those elements lead me to `won't fix`.

**Comment by apradhan12** [comment]
Thank you for the clarification! I really appreciate your thorough explanation.

From reading the discussions on the related issues and from this comment, I understand what you mean about this being a detail of a larger issue, and why you won't fix that larger issue for now. However, I think that the "`nolintlint` option to detect invalid cases and create a report for them" would still be valuable, as it avoids providing incorrect/confusing feedback to the user in this case. The fact that `nolintlint` provides an incorrect suggestion for cases like `// nolint: bob` (currently, the linter would suggest `//nolint: bob`) is a bug - the tool is providing a suggestion that is objectively wrong, and that bug can be fixed without needing to make a decision or make code changes addressing the larger issue of `nolintlint` allowing several kinds of directive syntax that are not officially permitted according to the Go directive syntax rules. Additionally, this bug, while certainly low priority, is confusing for users, especially because of the conflicting `go fmt` behavior that corrects such suggested fixes back to `// nolint: bob`. I encountered this issue and spent a while trying to figure out and understand what was going wrong, and the documentation is limited. Maybe it's an expectation that users will either already know Go directive syntax, or that, upon encountering this error, they will spend the time on their own to find and consult the Go documentation and learn the correct directive syntax, but if every developer who uses this tool and makes this kind of error has to go do that, that seems like an area in which the `nolintlint` tool could be improved. One of the goals of linters, as I see it, is to provide suggestions in an easy-to-understand and easy-to-apply manner to improve your code and make the code better conform to standards, and `nolintlint` is failing to meet that goal in this case, even though it is a relatively minor one in the grand scheme of things.

If it is deemed acceptable to implement an optional behavior in `nolintlint` to detect such invalid cases, I am more than willing to propose a design plan for this fix, and, if that design plan is approved, implement the behavior in code and go through the PR review process. I apologize for rushing into the PR stage initially - I now understand more fully that a behavior change of any kind for a tool that is used so widely and in so many different ways requires thorough discussion and approval by the community and the code owners. However, I do think there exists a path forward to fix this issue (within the scope of golangci-lint v1) that would be deemed acceptable by all parties. I don't yet know what that path is, but I'm happy to propose some ideas for discussion if you or other maintainers/code owners will be willing to review them. I won't contribute those ideas if this truly has zero chance of being fixed - I understand that certain features/bugfixes may not be considered worth it due to the existence of other higher-priority items, even though this is a bug that has real user impact - but I personally believe this is worth fixing, as someone who uses this library as a developer, and I think one of the assets of the open-source community is that anyone can contribute to help fix something like this, within the framework established by the particular community/project. If this issue has been worth the attention you've given it thus far to explain the context, I implore you to see the value in seeing this issue through to a fix.

As an aside, I also really appreciate your attentiveness to these issues and dedication to the `golangci-lint` community more broadly. As I said in the PR discussion, I'm new to the concept of contributing to open source, and I'm new to the `golangci-lint` community, but `golangci-lint` has been a hugely helpful tool in enforcing lint standards in my company's codebase, and I recognize the huge amount of volunteer work that you and others contribute to that community. I absolutely respect the dedication you put into this, and by putting up this issue and the aforementioned PR, I do not intend whatsoever to go against the direction of the community or to distract from more important issues, but rather, to help the community and improve the `golangci-lint` project in a very small way for others in the future. I feel that this issue is important enough to me to try to push to find a fix that will be suitable for the community and to help out with that process as much as I can, and I hope you will understand where I'm coming from and work with me to identify the best way to address this issue.

**Comment by ldez** [comment]
Sorry, but do you generate your answers with an AI?

You are writing a book with too many details each time, it's a bit discouraging to read.

If you are just a person who likes to write very long texts, don't take it badly, but I don't need an explanation about things that I already know, so you can go straight to the goal.

If you are using an AI, please generate a short answer.

**Comment by apradhan12** [comment]
Hi @ldez,

No, I'm not using AI. Please consider fixing this bug. I can propose a design plan for a "`nolintlint` option to detect the invalid cases and create a report for them" as you suggested, but I will only do so if you actually will seriously consider it and not just dismiss it immediately. It's been a bit frustrating that you respond immediately but ignore most of the content and dismiss what I'm saying - if you have time to respond immediately, you must surely have time to read more thoroughly and respond a little less promptly in exchange. Also, a more accurate title for this issue to convey the desired goal is: "Provide more helpful/specific warnings for nolint directives with invalid spacing."

**Comment by ldez** [comment]
> No, I'm not using AI.

I didn't say you are an AI, but that you are using an AI, not the same thing.

> It's been a bit frustrating that you respond immediately but ignore most of the content and dismiss what I'm saying

I think you have completely missed the meaning of my previous comment.

> if you have time to respond immediately, you must surely have time to read more thoroughly and respond a little less promptly in exchange.

You are frustrated, I can understand but your answer IS NOT ACCEPTABLE.
First warning.

**Comment by ldez** [comment]
We will not change `nolint` directives to follow strictly the convention.
`nolint` directives should be considered frozen with their imperfections.

I already said [here](https://github.com/golangci/golangci-lint/issues/4561#issuecomment-2015951238) the only acceptable possibility.

**Comment by apradhan12** [comment]
I apologize for what I said earlier.

Can I create a proposal for the design of the acceptable possibility you mentioned, i.e., "create a nolintlint option to DETECT the invalid cases and create a report for them (a user should be able to enable/disable this behavior to avoid usage of exclude about them)"? If the proposal were accepted, I would be willing to implement this design as well, if desired.
